src/engine/BoundsExplainer.cpp

```diff
@@ -0,0 +1,270 @@
+ ** Top contributors (to current version):
+ **   Omri Isac, Guy Katz
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2022 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+ **/
+
+#include "BoundsExplainer.h"
+
+BoundsExplainer::BoundsExplainer( const unsigned varsNum, const unsigned rowsNum )
+	:_varsNum( varsNum )
+	,_rowsNum( rowsNum )
+	,_upperBoundsExplanations( _varsNum, std::vector<double>( 0 ) )
+	,_lowerBoundsExplanations( _varsNum,  std::vector<double>( 0 ) )
+{
+}
+
+BoundsExplainer::~BoundsExplainer()
+{
+	for ( unsigned i = 0; i < _varsNum; ++i )
+	{
+		_upperBoundsExplanations[i].clear();
+		_lowerBoundsExplanations[i].clear();
+	}
+
+	_upperBoundsExplanations.clear();
+	_lowerBoundsExplanations.clear();
+}
+
+unsigned BoundsExplainer::getRowsNum() const
+{
+	return _rowsNum;
+}
+
+unsigned BoundsExplainer::getVarsNum() const
+{
+	return _varsNum;
+}
+
+BoundsExplainer& BoundsExplainer::operator=( const BoundsExplainer& other )
+{
+	if ( this == &other )
+		return *this;
+
+	if ( _varsNum != other._varsNum )
+	{
+		_upperBoundsExplanations.resize( other._varsNum );
+		_upperBoundsExplanations.shrink_to_fit();
+
+		_lowerBoundsExplanations.resize( other._varsNum );
+		_lowerBoundsExplanations.shrink_to_fit();
+	}
+
+	_rowsNum = other._rowsNum;
+	_varsNum = other._varsNum;
+
+	for ( unsigned i = 0; i < _varsNum; ++i )
+	{
+		if ( _upperBoundsExplanations[i].size() != other._upperBoundsExplanations[i].size() )
+		{
+			_upperBoundsExplanations[i].clear();
+			_upperBoundsExplanations[i].resize( other._upperBoundsExplanations[i].size() );
+			_upperBoundsExplanations[i].shrink_to_fit();
+		}
+
+		if ( _lowerBoundsExplanations[i].size() != other._lowerBoundsExplanations[i].size() )
+		{
+			_lowerBoundsExplanations[i].clear();
+			_lowerBoundsExplanations[i].resize( other._lowerBoundsExplanations[i].size() );
+			_lowerBoundsExplanations[i].shrink_to_fit();
+		}
+
+		std::copy( other._upperBoundsExplanations[i].begin(), other._upperBoundsExplanations[i].end(), _upperBoundsExplanations[i].begin() );
+		std::copy( other._lowerBoundsExplanations[i].begin(), other._lowerBoundsExplanations[i].end(), _lowerBoundsExplanations[i].begin() );
+	}
+
+	return *this;
+}
+
+const std::vector<double>& BoundsExplainer::getExplanation( const unsigned var, const bool isUpper )
+{
+	ASSERT ( var < _varsNum );
+	return isUpper ? _upperBoundsExplanations[var] : _lowerBoundsExplanations[var];
+}
+
+void BoundsExplainer::updateBoundExplanation( const TableauRow& row, const bool isUpper )
+{
+	if ( !row._size )
+		return;
+	updateBoundExplanation( row, isUpper, row._lhs );
+}
+
+void BoundsExplainer::updateBoundExplanation( const TableauRow& row, const bool isUpper, const unsigned var )
```

Comment:
const unsigned var is not needed; we only use const for things passed by pointers or by reference, not by value